Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+import Foundation
+import class PusherPlatform.PPRequestOptions
+import enum PusherPlatform.HTTPMethod
+
+typealias SubscribeHandler = (VoidResult) -> Void
+
+protocol SubscriptionDelegate: AnyObject {
+    func subscription(_ subscription: Subscription, didReceiveEventWithJsonData jsonData: Data)
+    func subscription(_ subscription: Subscription, didReceiveError error: Error)
+}
+
+protocol Subscription {
+    func subscribe(completion: @escaping SubscribeHandler)
+    func unsubscribe()
+}
+
+class ConcreteSubscription: Subscription {
+    
+    enum State {
+        case notSubscribed
+        case subscribingStageOne(instanceWrapper: InstanceWrapper, completions: [SubscribeHandler])
+        case subscribingStageTwo(instanceWrapper: InstanceWrapper, resumableSubscription: ResumableSubscription, completions: [SubscribeHandler])
+        case subscribed(instanceWrapper: InstanceWrapper, resumableSubscription: ResumableSubscription)
+    }
+    
+    typealias Dependencies = HasInstanceWrapperFactory
+    
+    let subscriptionType: SubscriptionType // Internal `get` aids testing
+    private let dependencies: Dependencies
+    weak var delegate: SubscriptionDelegate?
+    
+    private(set) var state: State = .notSubscribed // Internal `get` aids testing
+    
+    init(subscriptionType: SubscriptionType, dependencies: Dependencies, delegate: SubscriptionDelegate?) {
```

Comment:
If the `delegate` is optional why do we pass it in the initializer?

Reply:
A `Subscription` is pretty much useless without a `delegate` so IMO it makes sense for the initialiser to effectively prompt the user to pass/set a `delegate`.  In-fact I've now updated the code to make the argument non-optional.  The property has to be remain optional because its defined as `weak`.